ansible_rulebook/websocket.py

```diff
@@ -12,123 +12,254 @@
+                if new_token:
+                    extra_headers["Authorization"] = f"Bearer {new_token}"
+                else:
+                    raise
+            else:
+                # Sleep and retry implemention duplicated from
+                # websockets.lagacy.client.Client
 
+                # Add a random initial delay between 0 and 5 seconds.
+                # See 7.2.3. Recovering from Abnormal Closure in RFC 6544.
+                if backoff_delay == BACKOFF_MIN:
+                    initial_delay = random.random() * BACKOFF_INITIAL
+                    logger.info(
+                        "! connect failed; reconnecting in %.1f seconds",
+                        initial_delay,
+                        exc_info=True,
+                    )
+                    await asyncio.sleep(initial_delay)
+                else:
+                    logger.info(
+                        "! connect failed again; retrying in %d seconds",
+                        int(backoff_delay),
+                        exc_info=True,
+                    )
+                    await asyncio.sleep(int(backoff_delay))
+                # Increase delay with truncated exponential backoff.
+                backoff_delay = backoff_delay * BACKOFF_FACTOR
+                backoff_delay = min(backoff_delay, BACKOFF_MAX)
+                continue
+        else:
+            # Connection succeeded - reset backoff delay
+            backoff_delay = BACKOFF_MIN
 
-async def send_event_log_to_websocket(
-    event_log, websocket_address, websocket_ssl_verify
-):
-    logger.info("feedback websocket %s connecting", websocket_address)
-    event = None
-    async for websocket in websockets.connect(
-        websocket_address,
-        logger=logger,
-        ssl=_sslcontext(websocket_address, websocket_ssl_verify),
-    ):
-        logger.info("feedback websocket %s connected", websocket_address)
-        try:
-            if event:
-                logger.info("Resending last event...")
-                await websocket.send(json.dumps(event))
-                event = None
-
-            while True:
-                event = await event_log.get()
-                logger.debug(f"Event received, {event}")
-
-                if event == dict(type="Exit"):
-                    logger.info("Exiting feedback websocket task")
-                    return
-
-                await websocket.send(json.dumps(event))
-                event = None
-        except websockets.exceptions.ConnectionClosed:
-            logger.warning(
-                "feedback websocket %s connection closed, will retry...",
-                websocket_address,
+    return result
+
+
+async def request_workload(
+    activation_id: str,
+    websocket_address: str,
+    websocket_ssl_verify: str,
+    websocket_token_address: str,
+    websocket_access_token: str,
+    websocket_refresh_token: str,
+) -> StartupArgs:
+    return await _connect_websocket(
+        websocket_address=websocket_address,
```

Comment:
@bzwei would it make sense to move all these to settings https://github.com/ansible/ansible-rulebook/blob/8743f53a1a7bf3eb24dfbc4cbe03c1cd1fad66da/ansible_rulebook/conf.py#L23 and use it from there we don't have to keep passing it around to every function, they are part of the settings and constant once the process starts, the only thing changing would be the access_token.

Reply:
@bzwei We were planning on getting rid of the StartupArgs and just use settings every where and start using dynaconf.

Reply:
I just amend the existing cli arguments for websocket. Sure I can try to move all websocket options to settings.